htlcswitch/hop/payload.go

```diff
@@ -53,15 +88,24 @@ func NewPayloadFromReader(r io.Reader) (*Payload, error) {
 		return nil, err
 	}
 
-	_, err = tlvStream.DecodeWithParsedTypes(r)
+	parsedTypes, err := tlvStream.DecodeWithParsedTypes(r)
+	if err != nil {
+		return nil, err
+	}
+
+	nextHop := lnwire.NewShortChanIDFromInt(cid)
+
+	// Validate whether the sender properly included or omitted tlv records
+	// in accordance with BOLT 04.
+	err = ValidateParsedPayloadTypes(parsedTypes, nextHop == Exit)
```

Comment:
Still there is something not ideal here. The function does the validation, but part of the validation (`nextHop == Exit`) happens outside of the function.

Reply:
`nextHop` is now passed directly